### PR TITLE
Add run-mode field to status endpoint

### LINKED ISSUE #2612: Add run-mode field to casper-node status endpoint.
This should announce what mode the node is running to not require inference based on other fields or RPC being up, etc.
      

-  linear_sync, running_sync, trie_store_check_hell, etc.

### PR DESCRIPTION
**Pull request by darthsiroftardis**
CHANGELOG:

* Added a run-mode to the status endpoint 

Closes #2612

### CODE REVIEW

File: node/CHANGELOG.md [hunk 1]
```diff
@@ -26,6 +26,7 @@ All notable changes to this project will be documented in this file.  The format
 * Add `verifiable_chunked_hash_activation` to the chainspec to specify the first era in which the new Merkle tree-based hashing scheme is used.
 * In addition to `consensus` and `deploy_requests`, the following values can now be controlled via the `[network.estimator_weights]` section in config: `gossip`, `finality_signatures`, `deploy_responses`, `block_requests`, `block_responses`, `trie_requests` and `trie_responses`.
 * Nodes will now also gossip deploys onwards while joining.
+* Added run-mode field to the status endpoint.
```
Comment (rafal-ch) [inline node/CHANGELOG.md]:
```suggestion
* Added run-mode field to the `/status` endpoint and the `info_get_status` JSON-RPC.
```

File: node/src/components/rest_server.rs [hunk 2]
```diff
@@ -47,7 +47,10 @@ use crate::{
     NodeRng,
 };
 
-use crate::{components::rpc_server::rpcs::docs::OPEN_RPC_SCHEMA, effect::requests::RestRequest};
+use crate::{
+    components::rpc_server::rpcs::docs::OPEN_RPC_SCHEMA, effect::requests::RestRequest,
+    types::NodeState,
+};
```
Comment (rafal-ch) [inline node/src/components/rest_server.rs]: I think we can merge all `crate::` imports like so:
```
use crate::{
    components::rpc_server::rpcs::docs::OPEN_RPC_SCHEMA,
    effect::{
        requests::{
            ChainspecLoaderRequest, ConsensusRequest, MetricsRequest, NetworkInfoRequest,
            RestRequest, StorageRequest,
        },
        EffectBuilder, EffectExt, Effects,
    },
    reactor::Finalize,
    types::{NodeState, StatusFeed},
    utils::{self, ListeningError},
    NodeRng,
};
```

File: node/src/components/rest_server.rs [hunk 3]
```diff
@@ -87,6 +90,8 @@ pub(crate) struct RestServer {
     server_join_handle: Option<JoinHandle<()>>,
     /// The instant at which the node has started.
     node_startup_instant: Instant,
+    /// The current state of operation
```
Comment (rafal-ch) [inline node/src/components/rest_server.rs]:
```suggestion
    /// The current state of operation.
```

File: node/src/types/status_feed.rs [hunk 4]
```diff
@@ -74,6 +76,17 @@ impl ChainspecInfo {
...
+#[derive(Clone, Copy, DataSize, Debug, Deserialize, Serialize, JsonSchema)]
+pub enum NodeState {
+    /// The node is currently in the fast syncing state.
+    FastSyncing,
+    /// The node is currently in the archival syncing state.
+    ArchivalSyncing,
+    /// The node is currently in the participating state.
+    Participating,
+}
+
```
Comment (rafal-ch) [inline node/src/types/status_feed.rs]: As soon as [this PR](https://github.com/casper-network/casper-node/pull/2704) is merged the names of the processes will be changed. Please see the description of the mentioned PR.

**Review by rafal-ch — CHANGES_REQUESTED** [review]
Looks good with some minor nits :+1: 

However, I'd like the https://github.com/casper-network/casper-node/pull/2704 to be merged first, because it'll impact the fields of the `NodeState` that is introduced in this PR.

**Comment by rafal-ch** [comment]
@darthsiroftardis The https://github.com/casper-network/casper-node/pull/2704 is now merged.

File: node/src/types/status_feed.rs [hunk 5]
```diff
@@ -74,6 +76,17 @@ impl ChainspecInfo {
     }
 }
 
+/// The various possible states of operation for the node.
+#[derive(Clone, Copy, DataSize, Debug, Deserialize, Serialize, JsonSchema)]
+pub enum NodeState {
+    /// The node is currently in the fast syncing state.
+    FastSyncing,
+    /// The node is currently in the archival syncing state.
+    ArchivalSyncing,
```
Comment (Fraser999) [inline node/src/types/status_feed.rs]:
```suggestion
    /// The node is currently in syncing to genesis.
    SyncingToGenesis,
```

Comment (darthsiroftardis) [reply node/src/types/status_feed.rs]: Applied suggestion

File: node/src/types/status_feed.rs (hunk 4 above)
Comment (darthsiroftardis) [reply node/src/types/status_feed.rs]: Renamed as requested

File: node/src/components/rest_server.rs (hunk 3 above)
Comment (darthsiroftardis) [reply node/src/components/rest_server.rs]: Applied suggestion

File: node/src/components/rest_server.rs (hunk 2 above)
Comment (darthsiroftardis) [reply node/src/components/rest_server.rs]: Applied suggestion

File: node/CHANGELOG.md (hunk 1 above)
Comment (darthsiroftardis) [reply node/CHANGELOG.md]: Applied suggestion

File: node/src/reactor/joiner.rs [hunk 6]
```diff
@@ -593,11 +593,17 @@ impl reactor::Reactor for Reactor {
             .protocol_config
             .verifiable_chunked_hash_activation;
         let protocol_version = &chainspec_loader.chainspec().protocol_config.version;
+        let node_state = if config.node.archival_sync {
```
Comment (goral09) [inline node/src/reactor/joiner.rs]: I'm surprised this is still being called `archival_sync`. Should probably be renamed.

**Review by goral09 — APPROVED** [review]

File: node/src/reactor/joiner.rs (hunk 6 above)
Comment (darthsiroftardis) [reply node/src/reactor/joiner.rs]: Yes, it was cause I didn't merge the changes from the related PR, I merged the changes

File: node/src/types/status_feed.rs [hunk 7]
```diff
@@ -74,6 +76,17 @@ impl ChainspecInfo {
     }
 }
 
+/// The various possible states of operation for the node.
+#[derive(Clone, Copy, DataSize, Debug, Deserialize, Serialize, JsonSchema)]
+pub enum NodeState {
+    /// The node is currently in the fast syncing state.
+    FastSyncing,
+    /// The node is currently in syncing to genesis.
```
Comment (rafal-ch) [inline node/src/types/status_feed.rs]:
```suggestion
    /// The node is currently in syncing to genesis state.
```

**Review by rafal-ch — APPROVED** [review]
LGTM :+1:

**Pull request merged**